portal/src/main/scala/cromwell/pipeline/service/AuthService.scala

```diff
@@ -0,0 +1,59 @@
+package cromwell.pipeline.service
+
+import java.nio.charset.StandardCharsets
+import java.time.Instant
+import java.util.UUID
+
+import com.typesafe.config.Config
+import cromwell.pipeline.datastorage.dao.repository.UserRepository
+import cromwell.pipeline.datastorage.dto.{Credentials, User, UserId}
+import cromwell.pipeline.utils.auth.AuthTokenContent
+import cromwell.pipeline.utils.transformation.MonadTransformers._
+import org.apache.commons.codec.digest.DigestUtils
+import pdi.jwt.algorithms.JwtHmacAlgorithm
+import pdi.jwt.{Jwt, JwtAlgorithm, JwtClaim}
+import play.api.libs.json.Json
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.Random
+
+class AuthService(config: Config, userRepository: UserRepository)(implicit executionContext: ExecutionContext) {
+
+  private val SecretKey = config.getString("auth.secretKey")
+  private val HmacAlgorithm = {
+    val hmacAlgorithm = config.getString("auth.hmacAlgorithm")
+    JwtAlgorithm.fromString(hmacAlgorithm).asInstanceOf[JwtHmacAlgorithm]
+  }
+  private val ExpirationTimeInSeconds = config.getLong("auth.expirationTimeInSeconds")
+
+  def signIn(credentials: Credentials): Future[Option[String]] = {
+    userRepository
+      .getUserByEmail(credentials.email)
+      .filterT(user => user.passwordHash == calculatePasswordHash(credentials.password, user.passwordSalt))
+      .mapT(user => getToken(AuthTokenContent(user.userId.value)))
+  }
+
+  def signUp(credentials: Credentials): Future[String] = {
+    val passwordSalt = Random.nextLong().toHexString
+    val passwordHash = calculatePasswordHash(credentials.password, passwordSalt)
+    val newUser = User(userId = UserId(UUID.randomUUID().toString),
+      email = credentials.email,
+      passwordSalt = Random.nextLong().toHexString,
+      passwordHash = passwordHash)
+
+    userRepository.addUser(newUser).map(userId => getToken(AuthTokenContent(userId.value)))
+  }
+
+  private def calculatePasswordHash(password: String, salt: String): String = {
+    DigestUtils.sha256Hex((password + salt).getBytes(StandardCharsets.UTF_8))
```

Comment:
Could we use `java` default solution for encoding algorithms?
```scala
import java.security.MessageDigest

val encoder = MessageDigest.getInstance("SHA-256")
val hash = new BigInteger(1, encoder.digest((password + salt).getBytes(StandardCharsets.UTF_8)))
val hexString = String.format("%032x", hash)
```

We may keep it somewhere in a utility class. I just don't think it is a good idea to use separate library for such a simple task.

Reply:
Thank you for your suggestion.
It has been implemented ;)